rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1559 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
+
+//! Client for Azure Device Registry operations.
+//!
+//! To use this client, the `azure_device_registry` feature must be enabled.
+
+use std::{collections::HashMap, sync::Arc, time::Duration};
+
+use azure_iot_operations_mqtt::interface::{AckToken, ManagedClient};
+use azure_iot_operations_protocol::application::ApplicationContext;
+use derive_builder::Builder;
+use tokio::sync::Notify;
+
+use crate::azure_device_registry::device_name_gen::adr_base_service::client as adr_name_gen;
+use crate::azure_device_registry::{
+    Asset, AssetStatus, AssetUpdateObservation, Device, DeviceStatus, DeviceUpdateObservation,
+    Error, ErrorKind,
+    device_name_gen::{
+        common_types::options::CommandInvokerOptionsBuilder,
+        common_types::options::TelemetryReceiverOptionsBuilder,
+    },
+};
+use crate::common::dispatcher::{DispatchError, Dispatcher};
+
+const DEVICE_NAME_TOPIC_TOKEN: &str = "deviceName";
+const DEVICE_NAME_RECEIVED_TOPIC_TOKEN: &str = "ex:deviceName";
+const INBOUND_ENDPOINT_NAME_TOPIC_TOKEN: &str = "inboundEndpointName";
+const INBOUND_ENDPOINT_NAME_RECEIVED_TOPIC_TOKEN: &str = "ex:inboundEndpointName";
+
+/// Options for the Azure Device Registry client.
+#[derive(Builder, Clone, Default)]
+#[builder(setter(into))]
+pub struct ClientOptions {
+    /// If true, update notifications are auto-acknowledged
+    #[builder(default = "true")]
+    notification_auto_ack: bool,
+}
+
+/// Azure Device Registry client implementation.
+#[derive(Clone)]
+pub struct Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync + 'static,
+    C::PubReceiver: Send + Sync,
+{
+    // general
+    shutdown_notifier: Arc<Notify>,
+    // device
+    get_device_command_invoker: Arc<adr_name_gen::GetDeviceCommandInvoker<C>>,
+    update_device_status_command_invoker: Arc<adr_name_gen::UpdateDeviceStatusCommandInvoker<C>>,
+    notify_on_device_update_command_invoker:
+        Arc<adr_name_gen::SetNotificationPreferenceForDeviceUpdatesCommandInvoker<C>>,
+    device_update_notification_dispatcher: Arc<Dispatcher<(Device, Option<AckToken>)>>,
+    // asset
+    get_asset_command_invoker: Arc<adr_name_gen::GetAssetCommandInvoker<C>>,
+    update_asset_status_command_invoker: Arc<adr_name_gen::UpdateAssetStatusCommandInvoker<C>>,
+    notify_on_asset_update_command_invoker:
+        Arc<adr_name_gen::SetNotificationPreferenceForAssetUpdatesCommandInvoker<C>>,
+    asset_update_notification_dispatcher: Arc<Dispatcher<(Asset, Option<AckToken>)>>,
+}
+
+impl<C> Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync + 'static,
+    C::PubReceiver: Send + Sync,
+{
+    // ~~~~~~~~~~~~~~~~~ General APIs ~~~~~~~~~~~~~~~~~~~~~
+    /// Create a new Azure Device Registry Client.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidClientId`](ErrorKind::InvalidClientId)
```

Comment:
if you haven't run `cargo doc` yet and verified that the output looks good/everything links up properly, definitely do that before finalizing the PR!